### ISSUE TITLE
Why only tune query and value in every attention block?

### ISSUE THREAD

**Issue by LightersWang** [JamesQFreeman/LoRA-ViT#5]
Hello @JamesQFreeman,

Thanks for your great code for easy re-implementation of LoRA in ViT. It's very useful for us to adapt to our own task.

However, I have observed that only the query and value weights in the attention block are being tuned. I'm curious about the rationale behind this. In the original LoRA paper, the authors recommend tuning all four weights in the attention block (query, key, value, and projection linear layer) for better performance. 

Is there a trade-off between computational efficiency and final performance that led to this decision? What's more, have you seen an improvement in results by including the key and projection linear layer weights for tuning in your own task?

Thank you, and I look forward to hearing from you.

**Comment by leoinnocenzi** [comment]
Hello,

After reviewing the [LoRA paper](http://arxiv.org/abs/2106.09685), it seems that the authors do not actually recommend using the q/k/v/o setting. Using the four weight types is doubling the size of the LoRA modules compared to a query/value setting, which can slows down training. 

The tradeoff is indeed using less parameters with a slightly higher rank to speed up the training process and encode more information to the LoRA modules. When looking at Table 6 in the paper, one can decide of if this tradeoff is worth it or not, as q/v and k/q/v/o have a ~0.1-0.4 difference of validation accuracy in favour of the four weights setting.  I think this is an important tradeoff, saving a lot of memory while having comparable performances!

From experience, it seems that adding the key and output weights to the LoRA do not have a high enough performance bump to justify a higher need in computational ressources. 

Good luck with your experiments!
